torch_xla/csrc/aten_cpu_fallback.cpp

```diff
@@ -1,34 +1,355 @@
+//      before crossing device borders for correctness.
+//
+void cuda_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack,
+                   bool error_on_views) {
+  auto& schema_args = op.schema().arguments();
+  const auto num_arguments = schema_args.size();
+  auto arguments = torch::jit::last(stack, num_arguments);
+  const auto arguments_begin = stack->size() - num_arguments;
+
+  std::vector<at::Tensor> tensor_args;
+  std::vector<int> tensor_args_indices;
+
+  std::vector<c10::List<at::Tensor>> tensorlist_args;
+  std::vector<int> tensorlist_args_indices;
+
+  std::vector<c10::IValue> tensorlist_cuda_args;
+
+  // This fallback only works if all XLA:CUDA tensor arguments are
+  // on the same CUDA device.
+  //
+  // We keep track of said device, so that after actually running
+  // the operation on PyTorch CUDA eager-mode, we synchronize the
+  // device.
+  //
+  // This variable is updated over the course of 'to_cuda' calls.
+  c10::DeviceIndex common_device = -1;
+
+  // Initialize CUDA device.
+  torch::utils::device_lazy_init(at::kCUDA);
+
+  // Step 1: Convert all non-CUDA tensor inputs into CUDA tensors
+  // and put them on the stack at the correct indices.
+  for (const auto idx : c10::irange(arguments.size())) {
+    const auto& ivalue = arguments[idx];
+    if (ivalue.isTensor()) {
+      tensor_args.push_back(ivalue.toTensor());
+      tensor_args_indices.push_back(idx);
+    } else if (ivalue.isTensorList()) {
+      // Note: we copy each TensorList argument to CUDA individually out of
+      // convenience, but XLA would benefit from materializing all tensor and
+      // TensorList args onto the CUDA at the same time. We can improve this if
+      // we need better perf for XLA's CUDA fallbacks.
+      tensorlist_args.push_back(ivalue.toTensorList());
+      tensorlist_args_indices.push_back(idx);
+      auto cuda_ivalue = c10::IValue(c10::List<at::Tensor>(
+          to_cuda(ivalue.toTensorList().vec(), &common_device)));
+      tensorlist_cuda_args.push_back(cuda_ivalue);
+      (*stack)[arguments_begin + idx] = std::move(cuda_ivalue);
+      tensorlist_args.push_back(ivalue.toTensorList());
+    } else if (ivalue.isOptionalTensorList()) {
+      auto opt_tensors = ivalue.toOptionalTensorList().vec();
+      std::vector<at::Tensor> need_convert_tensors;
+      std::vector<int> need_convert_tensors_index;
+      for (auto i : c10::irange(opt_tensors.size())) {
+        if (!opt_tensors[i].has_value() || !opt_tensors[i]->defined()) continue;
+        need_convert_tensors.push_back(opt_tensors[i].value());
+        need_convert_tensors_index.push_back(i);
+      }
+      auto cuda_tensors = to_cuda(need_convert_tensors, &common_device);
+      for (const auto i : c10::irange(need_convert_tensors_index.size())) {
+        auto idx = need_convert_tensors_index[i];
+        opt_tensors[idx] = cuda_tensors[i];
+      }
+      (*stack)[arguments_begin + idx] = c10::IValue(opt_tensors);
+    }
```

Comment:
In the cpu implementation, there is a 
```
else if (ivalue.isDevice()) {
      tgt_device = ivalue.toDevice();
      (*stack)[arguments_begin + idx] = c10::IValue(c10::Device(kCPU));
    }
```
why don't we need it?

Reply:
Right. I thought we didn't need it, since we would always be on XLA. But, I guess it's important to have it just to be safe.

Reply:
Done. Let me know what you think.